src/NexusMods.DataModel/Serializers/DiskStateTreeSerializer.cs

```diff
@@ -0,0 +1,114 @@
+using System.Buffers;
+using System.Runtime.InteropServices;
+using FlatSharp;
+using NexusMods.Abstractions.DiskState;
+using NexusMods.Abstractions.GameLocators;
+using NexusMods.Archives.Nx.Enums;
+using NexusMods.Archives.Nx.Utilities;
+using NexusMods.DataModel.Serializers.DiskStateTreeSchema;
+using NexusMods.Hashing.xxHash64;
+using NexusMods.MnemonicDB.Abstractions;
+using NexusMods.MnemonicDB.Storage;
+using NexusMods.Paths;
+using NexusMods.Paths.Trees.Traits;
+using Reloaded.Memory.Extensions;
+using File = NexusMods.DataModel.Serializers.DiskStateTreeSchema.File;
+
+namespace NexusMods.DataModel.Serializers;
+
+/// <summary>
+/// We want to store the disk state tree in the datom store, so we'll need to serialize it.
+/// Long term we want to redo this, as we're just shoving potentially MB of data in a value, and
+/// that's not really what MnemonicDB is designed for. It works for now and is fairly easy to fix
+/// in the future, perhaps with MnemonicDB adding support for large values.
+///
+/// Another long-term option would be to store the tree *as* a tree in the datom store, that would
+/// require more diffing logic, but would give us the ability to perform more complex audits.
+/// </summary>
+internal class DiskStateTreeSerializer : IValueSerializer<DiskStateTree>
+{
+    public int Compare(in ReadOnlySpan<byte> a, in ReadOnlySpan<byte> b)
+    {
+        return a.SequenceCompareTo(b);
+    }
+
+    public Type NativeType => typeof(DiskStateTree);
+    public Symbol UniqueId => Symbol.Intern<DiskStateTreeSerializer>();
+    
+    public DiskStateTree Read(ReadOnlySpan<byte> buffer)
+    {
+        var decompressedSize = MemoryMarshal.Read<uint>(buffer);
+        var compressedData = buffer.SliceFast(sizeof(uint));
+
+        Files files;
+
+        if (decompressedSize == buffer.Length - sizeof(uint))
+        {
+            files = Files.Serializer.Parse(compressedData.ToArray());
+        }
+        else
+        {
+            var decompressed = new byte[decompressedSize];
+            unsafe
+            {
+                fixed (byte* src = compressedData)
+                fixed (byte* dest = decompressed)
+                {
+                    Compression.Decompress(CompressionPreference.ZStandard, src, compressedData.Length,
+                        dest, (int)decompressedSize
+                    );
+                }
+            }
+            files = Files.Serializer.Parse(decompressed);
+        }
+        
+        var kvs = files.All
+            .Select(f => new KeyValuePair<GamePath, DiskStateEntry>(new GamePath(LocationId.From(f.LocationId), f.Path),
+                new DiskStateEntry
+                {
+                    Size = Size.From(f.Size),
+                    Hash = Hash.From(f.Hash),
+                    LastModified = DateTime.FromFileTimeUtc(f.LastWriteTime)
+                }
+            ));
```

Comment:
We can replace this LINQ query with a simple for-loop. This is faster and more efficient since it doesn't have unnecessary allocations here. The subsequent iteration of the `IEnumerable<>` in `AGamePathNodeTree` will also be faster because no enumerator has to be allocated for arrays.

```csharp
        var res = GC.AllocateUninitializedArray<KeyValuePair<GamePath, DiskStateEntry>>(files.All.Count);

        for (var i = 0; i < files.All.Count; i++)
        {
            var f = files.All[i];
            res[i] = new KeyValuePair<GamePath, DiskStateEntry>(
                new GamePath(LocationId.From(f.LocationId), f.Path),
                new DiskStateEntry
                {
                    Size = Size.From(f.Size),
                    Hash = Hash.From(f.Hash),
                    LastModified = DateTime.FromFileTimeUtc(f.LastWriteTime),
                }
            );
        }
```